### PR TITLE
Add iptables --hex-string support to firewall resource

### PR DESCRIPTION
**Pull request by alexconrey**
**TL;DR**
Add `string_hex` parameter to the `firewall` resource when using iptables. Hex matching is useful for packet marking among other things.

**Why?**
Blocking a matching hex string is very useful for various types of attacks. The real world use case for this (aka. my motivation) was to block DNS reflection attacks. While the proper fix to these issues is usually within the application itself, sometimes we need to mitigate the issue until code can be fixed. With this PR, the following becomes possible:

```
common::fw_rules:
  '002 mark dns ANY queries UDP':
    proto: 'udp'
    dport: 53
    string_algo: 'bm'
    string_from: 40
    string_hex: '|0000FF0001|'
    recent: 'set'
    rname: 'dnsanyquery'
  '002 mark dns ANY queries TCP':
    proto: 'tcp'
    dport: 53
    string_algo: 'bm'
    string_from: 52
    string_hex: '|0000FF0001|'
    recent: 'set'
    rname: 'dnsanyquery'
  '002 ratelimit dns ANY queries UDP':
    proto: 'udp'
    action: 'drop'
    dport: 53
    string_algo: 'bm'
    string_from: 40
    string_hex: '|0000FF0001|'
    recent: 'rcheck'
    rname: 'dnsanyquery'
    rseconds: 60
    rhitcount: 3
  '002 ratelimit dns ANY queries TCP':
    proto: 'tcp'
    action: 'drop'
    dport: 53
    string_algo: 'bm'
    string_from: 52
    string_hex: '|0000FF0001|'
    recent: 'rcheck'
    rname: 'dnsanyquery'
    rseconds: 60
    rhitcount: 3
``` 

Rebased against `master` (using my own branch, which i should have done in the first place) per @sanfrancrisko request here: https://github.com/puppetlabs/puppetlabs-firewall/pull/902#issuecomment-597591201

### CODE REVIEW

**Comment by carabasdaniel** [comment]
Hi @alexconrey 

Thank you for your contribution, this looks good.

**Comment by jkroepke** [comment]
@alexconrey @carabasdaniel 

Does this work for ip6tables, too?

**Comment by mnejedlo-tds** [comment]
Ip6tables provider as pull req. 923
